src/Microsoft.ML.Tokenizers/Tokenizer.cs

```diff
@@ -136,6 +137,76 @@ public int CountTokens(string text, bool considerSpecialTokens = true)
+        /// and the token count can be generated using the provided subtext offset and length.
+        /// </returns>
+        /// <exception cref="ArgumentNullException">The input text is null.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">The maximum token count must be greater than 0.</exception>
+        /// <remarks>
+        /// If the tokenizer has a normalizer, the returned text will be the normalized text. Otherwise the returned text will be the input text.
+        /// If <paramref name="fromStart"/> is true, the returned offset will be 0. Otherwise the returned offset will be the starting index of the subtext.
+        /// If the provided <paramref name="maxTokenCount"/> is greater than the token count of the input text, the returned length will be the length of the input text.
+        /// If the provided <paramref name="maxTokenCount"/> is smaller enough to hold smallest number of grouped Ids, the returned length will be 0 and returned TokenCount will be 0.
+        /// </remarks>
+        public (string Text, int Offset, int Length, int TokenCount) TrimWithinTokenLimit(string text, int maxTokenCount, bool fromStart = true, bool considerSpecialTokens = true)
+        {
+            if (text is null)
+            {
+                throw new ArgumentNullException(nameof(text));
+            }
+
+            if (maxTokenCount <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxTokenCount), "The max token count must be greater than 0.");
+            }
+
+            string normalized = Normalizer is not null ? Normalizer.Normalize(text) : text;
+            int idsCount = 0;
+
+            if (fromStart)
+            {
+                foreach (Split split in PreTokenizer.PreTokenize(normalized, considerSpecialTokens))
+                {
+                    int tokenCount = Model.CountTokens(split.TokenSpan, split.IsSpecialToken);
+
+                    if (tokenCount + idsCount > maxTokenCount)
+                    {
+                        return (normalized, 0, split.Offset.Index, idsCount);
+                    }
+
+                    idsCount += tokenCount;
+                }
+
+                return (normalized, 0, normalized.Length, idsCount);
+            }
+
+            // from end
+            Split[] splits = PreTokenizer.PreTokenize(normalized, considerSpecialTokens).ToArray();
+
+            for (int i = splits.Length - 1; i >= 0; i--)
+            {
+                Split split = splits[i];
+                int tokenCount = Model.CountTokens(split.TokenSpan, split.IsSpecialToken);
+
+                if (tokenCount + idsCount > maxTokenCount)
+                {
+                    return (normalized, split.Offset.Index + split.Offset.Length, normalized.Length - split.Offset.Index - split.Offset.Length, idsCount);
+                }
+
+                idsCount += tokenCount;
+            }
+
+            return (normalized, 0, normalized.Length, idsCount);
```

Comment:
This return statement is the same for both branches and could be deduped. You could dedup more, too, e.g.
```C#
string normalized = Normalizer is not null ? Normalizer.Normalize(text) : text;
int idsCount = 0;

IEnumerable<Split> splits = PreTokenizer.PreTokenize(normalized, considerSpecialTokens);
foreach (Split split in (fromStart ? splits : splits.Reverse())
{
    int tokenCount = Model.CountTokens(split.TokenSpan, split.IsSpecialToken);
    if (tokenCount > maxTokenCount - idsCount)
    {
        return fromStart ?
            (normalized, 0, split.Offset.Index, idsCount) :
            (normalized, split.Offset.Index + split.Offset.Length, normalized.Length - split.Offset.Index - split.Offset.Length, idsCount);
    }

    idsCount += tokenCount;
}

return (normalized, 0, normalized.Length, idsCount);
```